File: src/main/java/com/cxxcxx/zinecraft/core/structure/NationSettlements.java
package com.cxxcxx.zinecraft.core.structure;

import com.cxxcxx.zinecraft.api.world.structure.JigsawBuildingEntry;
import com.cxxcxx.zinecraft.api.world.structure.StructureCatalog;
import com.cxxcxx.zinecraft.core.Zinecraft;
import com.cxxcxx.zinecraft.core.biome.NationBiomes;
import com.mojang.datafixers.util.Pair;
import net.minecraft.resources.ResourceKey;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.levelgen.Heightmap.Types;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.List;

public final class NationSettlements {
  /**
   * Dense national settlement placement. The 17-chunk minimum center gap remains
   * larger than twice the 112-block Jigsaw radius, so neighbouring settlements
   * cannot overlap even when both assemblies reach their configured boundary.
   */
  static final int DENSE_SETTLEMENT_SPACING = 36;
  static final int DENSE_SETTLEMENT_SEPARATION = 16;
  static final int DENSE_SETTLEMENT_JIGSAW_DEPTH = 9;
  static final int DENSE_SETTLEMENT_MAX_DISTANCE = 112;
  @NotNull
  public static final NationSettlements INSTANCE = new NationSettlements();
  @NotNull
  public static final JigsawBuildingEntry AEGIR_SUBSEA_ENCLAVE = settlementWithDefaults(
      INSTANCE,
      "aegir_subsea_enclave",
      NationBiomes.AEGIR_ABYSSAL_SEA,
      41000001,
      "pressure_residence",
      "hydroponics_lab",
      "bathysphere_dock",
      "current_archive",
      DENSE_SETTLEMENT_SPACING,
      Types.OCEAN_FLOOR_WG,
      0,
      0.0F,
      1536,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry BOLIVAR_DOSSOLES_DISTRICT = settlementWithDefaults(
      INSTANCE,
      "bolivar_dossoles_district",
      NationBiomes.BOLIVAR_PLAIN,
      41000002,
      "canal_house",
      "beach_market",
      "race_workshop",
      "festival_hall",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry HIGASHI_SOKOGAWA_TOWN = settlementWithDefaults(
      INSTANCE,
      "higashi_sokogawa_town",
      NationBiomes.HIGASHI_SHADOW_RIFT,
      41000003,
      "machiya",
      "swordsmith",
      "tea_house",
      "magistrate_house",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry DURIN_IDEAL_CITY_BLOCK = settlementWithDefaults(
      INSTANCE,
      "durin_ideal_city_block",
      NationBiomes.DURIN_UNDERGROUND_GARDEN,
      41000004,
      "dome_apartment",
      "machine_shop",
      "arcade",
      "transit_station",
      DENSE_SETTLEMENT_SPACING,
      null,
      24,
      0.0F,
      1024,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry COLUMBIA_FRONTIER_TOWN = settlementWithDefaults(
      INSTANCE,
      "columbia_frontier_town",
      NationBiomes.COLUMBIA_SANDSTONE_WILDS,
      41000005,
      "prefab_house",
      "pioneer_lab",
      "logistics_depot",
      "sheriff_office",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry KAZIMIERZ_KNIGHT_BOROUGH = settlementWithDefaults(
      INSTANCE,
      "kazimierz_knight_borough",
      NationBiomes.KAZIMIERZ_KNIGHTLAND,
      41000006,
      "tenement",
      "armor_workshop",
      "sponsor_shop",
      "tournament_inn",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry KAZDEL_SARKAZ_SETTLEMENT = settlementWithDefaults(
      INSTANCE,
      "kazdel_sarkaz_settlement",
      NationBiomes.KAZDEL_SCARRED_WASTES,
      41000007,
      "canvas_house",
      "forge",
      "mercenary_lodge",
      "provision_store",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry LATERANO_MONASTERY_TOWN = settlementWithDefaults(
      INSTANCE,
      "laterano_monastery_town",
      NationBiomes.LATERANO_HOLY_FIELDS,
      41000008,
      "white_residence",
      "confectionery",
      "notary_office",
      "bell_chapel",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry LEITHANIEN_MUSIC_TOWN = settlementWithDefaults(
      INSTANCE,
      "leithanien_music_town",
      NationBiomes.LEITHANIEN_TWILIGHT_FOREST,
      41000009,
      "twilight_house",
      "instrument_workshop",
      "rehearsal_hall",
      "arts_academy",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry RIM_BILLITON_MINING_CAMP = settlementWithDefaults(
      INSTANCE,
      "rim_billiton_mining_camp",
      NationBiomes.RIM_BILLITON_MINING_BADLANDS,
      41000010,
      "miner_bunkhouse",
      "ore_workshop",
      "freight_depot",
      "canteen",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry MINOS_HEROIC_POLIS = settlementWithDefaults(
      INSTANCE,
      "minos_heroic_polis",
      NationBiomes.MINOS_SUNLIT_HILLS,
      41000011,
      "courtyard_house",
      "olive_market",
      "training_hall",
      "council_house",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry SARGON_OASIS_TOWN = settlementWithDefaults(
      INSTANCE,
      "sargon_oasis_town",
      NationBiomes.SARGON_ROCKY_DESERT,
      41000012,
      "adobe_house",
      "spice_market",
      "caravanserai",
      "well_house",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry SAMI_SNOWPRIEST_VILLAGE = settlementWithDefaults(
      INSTANCE,
      "sami_snowpriest_village",
      NationBiomes.SAMI_FROZEN_FOREST,
      41000013,
      "snow_lodge",
      "hunter_camp",
      "ritual_house",
      "supply_shed",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry VICTORIA_INDUSTRIAL_BOROUGH = settlementWithDefaults(
      INSTANCE,
      "victoria_industrial_borough",
      NationBiomes.VICTORIA_MISTY_HIGHLANDS,
      41000014,
      "brick_tenement",
      "steam_workshop",
      "rail_warehouse",
      "council_hall",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry URSUS_NORTHERN_TOWN = settlementWithDefaults(
      INSTANCE,
      "ursus_northern_town",
      NationBiomes.URSUS_FROZEN_STEPPE,
      41000015,
      "heated_house",
      "military_storehouse",
      "mine_office",
      "communal_hall",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry KJERAG_MOUNTAIN_VILLAGE = settlementWithDefaults(
      INSTANCE,
      "kjerag_mountain_village",
      NationBiomes.KJERAG_SNOWY_PEAKS,
      41000016,
      "stone_chalet",
      "tea_workshop",
      "caravan_post",
      "shrine_house",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry SIRACUSA_FAMILY_TOWN = settlementWithDefaults(
      INSTANCE,
      "siracusa_family_town",
      NationBiomes.SIRACUSA_RAINY_WOODLAND,
      41000017,
      "family_house",
      "trattoria",
      "tailor_shop",
      "meeting_hall",
      0,
      null,
      0,
      0.25F,
      896,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry YAN_SHANGSHU_TOWN = settlementWithDefaults(
      INSTANCE,
      "yan_shangshu_town",
      NationBiomes.YAN_MOUNTAIN_GROVE,
      41000018,
      "courtyard_residence",
      "tea_house",
      "artisan_workshop",
      "relay_office",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );
  @NotNull
  public static final JigsawBuildingEntry IBERIA_COASTAL_TOWN = settlementWithDefaults(
      INSTANCE,
      "iberia_coastal_town",
      NationBiomes.IBERIA_SALT_DELTA,
      41000019,
      "saltstone_house",
      "shipwright",
      "fish_market",
      "inquisitor_office",
      0,
      null,
      0,
      0.0F,
      1920,
      null
  );

  static {
    validateNationCoverage();
  }

  private NationSettlements() {
  }

  /**
   * Fails data generation early if a national biome loses its one-to-one settlement binding.
   */
  private static void validateNationCoverage() {
    List<JigsawBuildingEntry> settlements = List.of(
        AEGIR_SUBSEA_ENCLAVE, BOLIVAR_DOSSOLES_DISTRICT, HIGASHI_SOKOGAWA_TOWN,
        DURIN_IDEAL_CITY_BLOCK, COLUMBIA_FRONTIER_TOWN, KAZIMIERZ_KNIGHT_BOROUGH,
        KAZDEL_SARKAZ_SETTLEMENT, LATERANO_MONASTERY_TOWN, LEITHANIEN_MUSIC_TOWN,
        RIM_BILLITON_MINING_CAMP, MINOS_HEROIC_POLIS, SARGON_OASIS_TOWN,
        SAMI_SNOWPRIEST_VILLAGE, VICTORIA_INDUSTRIAL_BOROUGH, URSUS_NORTHERN_TOWN,
        KJERAG_MOUNTAIN_VILLAGE, SIRACUSA_FAMILY_TOWN, YAN_SHANGSHU_TOWN,
        IBERIA_COASTAL_TOWN
    );
    var biomes = new HashSet<ResourceKey<Biome>>();
    var structureIds = new HashSet<>();
    int fixedOriginCount = 0;
    for (JigsawBuildingEntry settlement : settlements) {
      if (settlement.getBiome() == null || !biomes.add(settlement.getBiome())) {
        throw new IllegalStateException("十九国群系必须各自唯一绑定一个本国聚落");
      }
      if (!structureIds.add(settlement.getStructureKey())) {
        throw new IllegalStateException("国家聚落结构 ID 不得重复");
      }
      if (settlement.getUnique()
          || settlement.getSize() != DENSE_SETTLEMENT_JIGSAW_DEPTH
          || settlement.getMaxDistanceFromCenter() != DENSE_SETTLEMENT_MAX_DISTANCE) {
        throw new IllegalStateException("国家普通聚落不得使用唯一地标放置，且必须遵守统一展开上限");
      }
      if (settlement.getFixedOrigin()) {
        fixedOriginCount++;
      } else if (settlement.getSpacing() != DENSE_SETTLEMENT_SPACING
          || settlement.getSeparation() != DENSE_SETTLEMENT_SEPARATION) {
        throw new IllegalStateException("外围国家聚落必须使用统一高密度随机散布参数");
      }
    }
    if (settlements.size() != 19 || biomes.size() != 19 || fixedOriginCount != 1) {
      throw new IllegalStateException("国家聚落覆盖必须恰好为 19 国");
    }
  }

  static JigsawBuildingEntry settlementWithDefaults(
      NationSettlements var0,
      String var1,
      ResourceKey var2,
      int var3,
      String var4,
      String var5,
      String var6,
      String var7,
      int var8,
      Types var9,
      int var10,
      float var11,
      int var12,
      Object var13
  ) {
    if ((var12 & 128) != 0) {
      var8 = DENSE_SETTLEMENT_SPACING;
    }

    if ((var12 & 256) != 0) {
      var9 = Types.WORLD_SURFACE_WG;
    }

    if ((var12 & 512) != 0) {
      var10 = 0;
    }

    if ((var12 & 1024) != 0) {
      var11 = 0.0F;
    }

    return var0.settlement(var1, var2, var3, var4, var5, var6, var7, var8, var9, var10, var11);
  }

  private final JigsawBuildingEntry settlement(
      String path,
      ResourceKey<Biome> biome,
      int salt,
      String first,
      String second,
      String third,
      String fourth,
      int spacing,
      Types heightmap,
      int startHeight,
      float removeVinesChance
  ) {
    StructureCatalog structureCatalog = Zinecraft.WORLDGEN.getStructures();
    String string = "nation_settlements/" + path;
    Pair[] pairs = new Pair[]{Pair.of(first, 4), Pair.of(second, 3), Pair.of(third, 2), Pair.of(fourth, 2)};
    var templates = com.cxxcxx.zinecraft.api.util.CollectionSupport.<String, Integer>linkedMapOf(pairs);
    if (spacing != DENSE_SETTLEMENT_SPACING) {
      throw new IllegalStateException("十九国普通聚落必须使用统一高密度 spacing: " + path);
    }
    if ("laterano_monastery_town".equals(path)) {
      return structureCatalog.fixedOriginSettlement(
          path, string, biome, salt, templates,
          DENSE_SETTLEMENT_JIGSAW_DEPTH, DENSE_SETTLEMENT_MAX_DISTANCE,
          heightmap, startHeight, removeVinesChance
      );
    }
    return structureCatalog.settlement(
        path, string, biome, salt, templates,
        spacing, DENSE_SETTLEMENT_SEPARATION,
        DENSE_SETTLEMENT_JIGSAW_DEPTH, DENSE_SETTLEMENT_MAX_DISTANCE,
        heightmap, startHeight, removeVinesChance
    );
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/structure/NationSettlements.kt
package com.cxxcxx.zinecraft.core.structure

import com.cxxcxx.zinecraft.core.Zinecraft
import com.cxxcxx.zinecraft.core.biome.NationBiomes
import net.minecraft.resources.ResourceKey
import net.minecraft.world.level.biome.Biome
import net.minecraft.world.level.levelgen.Heightmap

/** 十九个国家群系中可重复生成的大型 Jigsaw 城镇、村落和营地。 */
object NationSettlements {
  /** 阿戈尔海底聚居区：压力住宅、水培实验室、深潜码头与海流档案馆。 */
  val AEGIR_SUBSEA_ENCLAVE = settlement(
    "aegir_subsea_enclave", NationBiomes.AEGIR_ABYSSAL_SEA, 41000001,
    "pressure_residence", "hydroponics_lab", "bathysphere_dock", "current_archive",
    spacing = 64, heightmap = Heightmap.Types.OCEAN_FLOOR_WG
  )

  /** 玻利瓦尔多索雷斯街区：运河住宅、海滩市场、赛事工坊与庆典厅。 */
  val BOLIVAR_DOSSOLES_DISTRICT = settlement(
    "bolivar_dossoles_district", NationBiomes.BOLIVAR_PLAIN, 41000002,
    "canal_house", "beach_market", "race_workshop", "festival_hall"
  )

  /** 东国锁川町：町屋、锻刀铺、茶屋与奉行所。 */
  val HIGASHI_SOKOGAWA_TOWN = settlement(
    "higashi_sokogawa_town", NationBiomes.HIGASHI_SHADOW_RIFT, 41000003,
    "machiya", "swordsmith", "tea_house", "magistrate_house"
  )

  /** 杜林理想城街区：穹顶公寓、机械工坊、游戏厅与轨道站。 */
  val DURIN_IDEAL_CITY_BLOCK = settlement(
    "durin_ideal_city_block", NationBiomes.DURIN_UNDERGROUND_GARDEN, 41000004,
    "dome_apartment", "machine_shop", "arcade", "transit_station",
    spacing = 60, heightmap = null, startHeight = 24
  )

  /** 哥伦比亚拓荒镇：装配住宅、拓荒实验室、物流仓库与治安所。 */
  val COLUMBIA_FRONTIER_TOWN = settlement(
    "columbia_frontier_town", NationBiomes.COLUMBIA_SANDSTONE_WILDS, 41000005,
    "prefab_house", "pioneer_lab", "logistics_depot", "sheriff_office"
  )

  /** 卡西米尔骑士城区：公寓、甲胄工坊、赞助商店与竞赛旅店。 */
  val KAZIMIERZ_KNIGHT_BOROUGH = settlement(
    "kazimierz_knight_borough", NationBiomes.KAZIMIERZ_KNIGHTLAND, 41000006,
    "tenement", "armor_workshop", "sponsor_shop", "tournament_inn"
  )

  /** 卡兹戴尔萨卡兹聚落：帐屋、铸炉、佣兵会所与补给铺。 */
  val KAZDEL_SARKAZ_SETTLEMENT = settlement(
    "kazdel_sarkaz_settlement", NationBiomes.KAZDEL_SCARRED_WASTES, 41000007,
    "canvas_house", "forge", "mercenary_lodge", "provision_store"
  )

  /** 拉特兰修道院镇：白石住宅、甜品店、公证所与钟楼礼拜堂。 */
  val LATERANO_MONASTERY_TOWN = settlement(
    "laterano_monastery_town", NationBiomes.LATERANO_HOLY_FIELDS, 41000008,
    "white_residence", "confectionery", "notary_office", "bell_chapel"
  )

  /** 莱塔尼亚音乐镇：暮色住宅、乐器工坊、排练厅与艺术学院。 */
  val LEITHANIEN_MUSIC_TOWN = settlement(
    "leithanien_music_town", NationBiomes.LEITHANIEN_TWILIGHT_FOREST, 41000009,
    "twilight_house", "instrument_workshop", "rehearsal_hall", "arts_academy"
  )

  /** 雷姆必拓矿业营地：矿工宿舍、矿石工坊、货运站与食堂。 */
  val RIM_BILLITON_MINING_CAMP = settlement(
    "rim_billiton_mining_camp", NationBiomes.RIM_BILLITON_MINING_BADLANDS, 41000010,
    "miner_bunkhouse", "ore_workshop", "freight_depot", "canteen"
  )

  /** 米诺斯英雄城邦：院落住宅、橄榄市场、训练厅与议事厅。 */
  val MINOS_HEROIC_POLIS = settlement(
    "minos_heroic_polis", NationBiomes.MINOS_SUNLIT_HILLS, 41000011,
    "courtyard_house", "olive_market", "training_hall", "council_house"
  )

  /** 萨尔贡绿洲镇：土坯住宅、香料市场、商旅驿站与水井房。 */
  val SARGON_OASIS_TOWN = settlement(
    "sargon_oasis_town", NationBiomes.SARGON_ROCKY_DESERT, 41000012,
    "adobe_house", "spice_market", "caravanserai", "well_house"
  )

  /** 萨米雪祀村：雪屋、猎人营地、仪式屋与补给棚。 */
  val SAMI_SNOWPRIEST_VILLAGE = settlement(
    "sami_snowpriest_village", NationBiomes.SAMI_FROZEN_FOREST, 41000013,
    "snow_lodge", "hunter_camp", "ritual_house", "supply_shed"
  )

  /** 维多利亚工业城区：砖砌公寓、蒸汽工坊、铁路仓库与市政厅。 */
  val VICTORIA_INDUSTRIAL_BOROUGH = settlement(
    "victoria_industrial_borough", NationBiomes.VICTORIA_MISTY_HIGHLANDS, 41000014,
    "brick_tenement", "steam_workshop", "rail_warehouse", "council_hall"
  )

  /** 乌萨斯北方城镇：保温住宅、军需仓库、矿务所与公共大厅。 */
  val URSUS_NORTHERN_TOWN = settlement(
    "ursus_northern_town", NationBiomes.URSUS_FROZEN_STEPPE, 41000015,
    "heated_house", "military_storehouse", "mine_office", "communal_hall"
  )

  /** 谢拉格山村：石木民居、茶坊、商队驿站与祭祀屋。 */
  val KJERAG_MOUNTAIN_VILLAGE = settlement(
    "kjerag_mountain_village", NationBiomes.KJERAG_SNOWY_PEAKS, 41000016,
    "stone_chalet", "tea_workshop", "caravan_post", "shrine_house"
  )

  /** 叙拉古家族镇：家族住宅、餐馆、裁缝铺与议事厅。 */
  val SIRACUSA_FAMILY_TOWN = settlement(
    "siracusa_family_town", NationBiomes.SIRACUSA_RAINY_WOODLAND, 41000017,
    "family_house", "trattoria", "tailor_shop", "meeting_hall",
    removeVinesChance = 0.25f
  )

  /** 炎国尚蜀山城：院落住宅、茶馆、百工坊与驿站。 */
  val YAN_SHANGSHU_TOWN = settlement(
    "yan_shangshu_town", NationBiomes.YAN_MOUNTAIN_GROVE, 41000018,
    "courtyard_residence", "tea_house", "artisan_workshop", "relay_office"
  )

  /** 伊比利亚海岸镇：盐石住宅、造船坊、鱼市与审判庭办事处。 */
  val IBERIA_COASTAL_TOWN = settlement(
    "iberia_coastal_town", NationBiomes.IBERIA_SALT_DELTA, 41000019,
    "saltstone_house", "shipwright", "fish_market", "inquisitor_office"
  )

  private fun settlement(
    path: String,
    biome: ResourceKey<Biome>,
    salt: Int,
    first: String,
    second: String,
    third: String,
    fourth: String,
    spacing: Int = 52,
    heightmap: Heightmap.Types? = Heightmap.Types.WORLD_SURFACE_WG,
    startHeight: Int = 0,
    removeVinesChance: Float = 0f
  ) = Zinecraft.STRUCTURES.settlement(
    path = path,
    templateRoot = "nation_settlements/$path",
    biome = biome,
    salt = salt,
    buildingTemplates = linkedMapOf(first to 4, second to 3, third to 2, fourth to 2),
    spacing = spacing,
    separation = 24,
    size = 7,
    maxDistanceFromCenter = 112,
    heightmap = heightmap,
    startHeight = startHeight,
    removeVinesChance = removeVinesChance
  )
}
